Generate signed and unsigned rpm fixtures.

emmawu opened the pull request:
including,
  unsigned rpm fixtures
  signed srpm fixtures
  unsigned srpm fixtures
  unsigned drpm fixtures

Ichimonji10 commented on README.rst:
```diff
@@ -50,6 +56,12 @@ are omitted.
+    The ``createrepo`` and ``modifyrepo`` utilities must be available.
+
+``fixtures/srpm-unsigned``
+    The ``createrepo``, ``modifyrepo`` and ``rpmsign`` utilities must be available.
```
There is no such utility as `rpmsign`. Do you mean `rpm-sign`?

Ichimonji10 commented on Makefile:
```diff
@@ -20,21 +28,40 @@ all: fixtures
 	$(warning The `all` target is deprecated. Use `fixtures` instead.)
 
 fixtures: fixtures/docker \
+	fixtures/drpm-unsigned \
```
Nitpick: all of the other lines below use spaces, and using a tab character here is out-of-place.

Ichimonji10 commented on README.rst:
```diff
@@ -34,9 +34,15 @@ are omitted.
     The ``createrepo`` and ``modifyrepo`` utilities must be available.
 
+``fixtures/rpm-unsigned``
+    The ``createrepo``, ``modifyrepo`` and ``rpm-sign`` utilities must be available.
```
There is no executable named `rpm-sign`.  However, there is an executable named `rpmsign`. Do you mean that? (Confusingly, the `rpmsign` utility is provided by the `rpm-sign` package.)

emmawu replied:
I was thinking whether I should put package name or utility name there.. Later I found it should be the utility name, and updated descriptions in .sh files. But I forgot to change the name in README as well :(

Ichimonji10 replied:
:+1:

Yeah, it's an interesting choice to make. In my opinion, it's better to write the utility name, because that tends to be more stable across different distributions and through time.

Ichimonji10 commented on rpm/del-fixtures-sign.sh:
```diff
@@ -0,0 +1,44 @@
+fi
+
+# Delete signatures from packages in target_dir recursively
+rpmsign --delsign `find ${target_dir} -print | grep -i '.*[.][sd]*rpm'`
```
Do not use backticks. They're deprecated in favor of `$()`, not least because they cannot be nested.

Ichimonji10 commented on rpm/del-fixtures-sign.sh:
```diff
@@ -0,0 +1,44 @@
+fmt <<EOF
+usage: del-fixtures-sign.sh <target_dir>
+
+Delete signatures of all packages recursively in <target_dir>
```
What are the "all packages" referred to by this sentence? All files ending in `.rpm`, `.drpm` and `.srpm`? Please clarify.

emmawu replied:
"all packages" means all rpm/srpm/drpm files. Updated the description.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,160 @@
+for (( i=0; i<${rpms_len}-1; i++ ));
+do
+	old_package=${found_rpms[i]}
+	new_package=${found_rpms[i+1]}
```
Tabs are pure evil. :wink:

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,160 @@
+get_rpm_name() {
+    local filename="$(basename "${1}")"
+    IFS=$'\n' local parts=($(echo "${filename}" | tr - "\n"))
+    local end=`expr ${#parts[@]} - 2`
```
Backticks are deprecated and not part of the POSIX standard.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,160 @@
+set -euo pipefail
+
+#------------------------------------------------------------------------------
+# Helper Functions
```
Nitpick: either drop this headline, or add a "business logic" headline below. I'd advocate for the former.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,160 @@
+    fi
+done
+
+readonly num_needed=2
```
This is a single-use variable. Interpolate its value into the code below and drop this variable, or use it in the string error message.

emmawu replied:
Thanks. Added the variable to the error message.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,160 @@
+
+	delta_package="${new_name}-${old_version}-${old_release}_${new_version}-${new_release}.${new_arch}.drpm"
+
+	makedeltarpm ${old_package} ${new_package} "${output_dir}"/drpms/"${delta_package}"
```
Quote all variables to prevent word splitting issues. Also, limit lines to 80 characters wide if reasonably possible.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,160 @@
+    fi
+
+	old_arch=$(get_rpm_arch "${old_package}")
+	new_arch=$(get_rpm_arch "${new_package}")
```
Quote all variable expansions to prevent word splitting issues.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,160 @@
+done
+
+# Generate DRPMS fixtures from RPMS and DRPMS
+createrepo --checksum sha256 --delta ${output_dir}
```
Quote all variable expansions to prevent word splitting issues. (Try cloning all of your repositories into a directory named `my code`, sometime, and see how many things break. It's astounding.)

emmawu replied:
Thanks for reminding. I have updated and tested the script in a directory of which full path has spaces..It should be OK now.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,160 @@
+#------------------------------------------------------------------------------
+
+# Fetch output_dir from user.
+if [ "$#" -lt 2 ]; then
```
So... it's valid to provide two arguments, yet `package_name="${3}"` below?

Ichimonji10 commented:
This generally looks pretty good. However, my one remaining concern is the size of the yum DRPM files. At 1MB each, they add about 10MB to the size of the repository. That's OK, but not ideal, simply because of how much time it adds to repository clone times, and it'll compound if we ever decide to use different RPMs. Will that add another 10MB?

I should point out that I have a personal stake in this. I sometimes work from locations with slow connections that run at 1 Mbit/s or less.

Ichimonji10 commented on rpm/del-fixtures-sign.sh:
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env sh
```
Please declare the script as being a bash script, not a sh script. Some systems will ignore the request and start a bash shell anyway, but that's not something to count on.

Ichimonji10 commented on rpm/del-fixtures-sign.sh:
```diff
@@ -0,0 +1,41 @@
+fi
+
+# Delete signatures from packages in target_dir recursively
+find "${target_dir}" -name "*rpm" -exec rpmsign --delsign {} \;
```
`find`'s `-exec` argument frequently masks non-zero return values. Use `xargs` instead. It's better at complaining when thing go wrong. Also, `*rpm` will catch files such as `not_an_rpm`. Use `\( -name 'pattern1' -o -name 'pattern2' \)`.

```bash
find "${target_dir}" \( -name '*.rpm' -o -name '*.drpm' -o -name '*.srpm' \) \
    -print0 | xargs -0 rpmsign --delsign
```

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env sh
```
Declare this script as a bash script, not an sh script.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+readonly num_needed=2
+rpms_len=${#found_rpms[@]}
+if [ ${rpms_len}  -lt ${num_needed} ]; then
+    echo 1>&2 'Error: Need at least ${num_needed} RPMS. But found ${rpms_len}'
```
Variables are interpolated into double-quoted strings, but not single-quoted strings. This will literally print the following message:

    Error: Need at least ${num_needed} RPMS. But found ${rpms_len}

Ichimonji10 commented on rpm/del-fixtures-sign.sh:
```diff
@@ -0,0 +1,41 @@
+    show_help 1>&2
+    exit 1
+else
+    target_dir="$(realpath --canonicalize-missing "${1}")"
```
`--canonicalize-missing` is an inappropriate option to use here. From `realpath`'s man page:

```
       Print the resolved absolute file name; all but the last component must exist

       -e, --canonicalize-existing
              all components of the path must exist

       -m, --canonicalize-missing
              no path components need exist or be a directory
```

`--canonicalize-existing` should be used.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+get_rpm_name() {
+    local filename="$(basename "${1}")"
+    IFS=$'\n' local parts=($(echo "${filename}" | tr - "\n"))
+    local end=$(expr ${#parts[@]} - 2)
```
`expr` is ancient. Use `$(( ))` for math.

```bash
end=$(( ${#parts[@]} - 2 ))
```

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+    exit 1
+else
+    output_dir="$(realpath --canonicalize-missing "${1}")"
+    assets_dir="$(realpath "${2}")"
```
Consider using `--canonicalize-existing`.

emmawu replied:
I think this one should use option '--canonicalize-missing'. Unlike del-fixtures-sign.sh,  ${output_dir}" is not expected to exist before running gen-fixtures-delta.sh. It's removed by `make clean`. And gen-fixtures-delta.sh will create this directory ${output_dir}. 
This is also true to other scripts except del-fixture-sign.sh

Ichimonji10 replied:
Yes, `$output_dir` should use `--canonicalize-existing`. But I'm looking at `$assets_dir`.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+    arch_2="$(get_rpm_arch "${pkg_2}")"
+
+    if [ "${arch_1}" != "${arch_2}" ]; then
+        echo 1>&2 'Error: Old and new packages need has same arch'
```
Does the `makedeltarpm` tool perform this check? If so, let's omit this check and let that tool do error checking for us. If not, how about print this message instead?

```bash
    arch_1="$(get_rpm_arch "${pkg_1}")"
    arch_2="$(get_rpm_arch "${pkg_2}")"
    if [ "${arch_1}" != "${arch_2}" ]; then
        fmt 1>&2 <<EOF
Error: Old and new packages must have the same architecture, but are different.
Package architectures are ${arch_1} and ${arch_2}.
EOF
        exit 1
    fi
```

emmawu replied:
makedeltarpm doesn't do the check. It will take whatever rpm packages and generate the drpm package in the name specified in the 3rd argument.
Thanks. I have updated the error message.

Ichimonji10 replied:
Thanks for the clarification. :smile:

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+    name_2="$(get_rpm_name "${pkg_2}")"
+
+    if [ "${name_1}" != "${name_2}" ]; then
+        echo 1>&2 'Error: Packages need has same name'
```
Does the `makedeltarpm` tool perform this check? If so, let's omit this check and let that tool do error checking for us. If not, how about printing this message instead?

```bash
    name_1="$(get_rpm_name "${pkg_1}")"
    name_2="$(get_rpm_name "${pkg_2}")"
    if [ "${name_1}" != "${name_2}" ]; then
        fmt 1>&2 <<EOF
Error: Old and new packages must have the same name, but are different. Package
names are ${name_1} and ${name_2}.
EOF
        exit 1
    fi
```

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+cp -t "${output_dir}" "${assets_dir}/"*.rpm
+
+# Filter and sort all required RPMs.
+rpms="${output_dir}"/"${package_name}"*.rpm
```
This will literally assign the string `${output_dir}"/"${package_name}"*.rpm` to the `rpms` variable. Here's a simple example:

```bash
touch myfoo mybar
var=my*
echo "$var"  # my*
echo $var  # myfoo myvar
```

Also, glob expansion produces an array, so it'd be good to use array syntax if an array is expected.

Nitpick: A benefit of the `${...}` variable name syntax is that it's easy to find variables even in complicated strings. In my opinion, `"${foo}/${bar}"` is easier to read than `"${foo}"/"${bar}"`.

Try this instead:

```bash
rpms=( "$output_dir}/${package_name}"*.rpm )
```

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+
+readonly num_needed=2
+rpms_len=${#found_rpms[@]}
+if [ ${rpms_len}  -lt ${num_needed} ]; then
```
Use double quotes to prevent word expansion issues.

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+
+# Make DRPMS from RPMS
+# DRPMS is generated from all neighbouring RPMS in filtered RPMS list
+for (( i=0; i<${rpms_len}-1; i++ ));
```
This can be written as:

```bash
for (( i=0; i < rpms_len - 1; i++ )); do
```

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+    local filename="$(basename "${1}")"
+    IFS=$'\n' local parts=($(echo "${filename}" | tr - "\n"))
+    local end=$(expr ${#parts[@]} - 2)
+    for ((i=0; i<${end}; i++)); do
```
This can be written as:

```bash
for (( i=0; i < end; i++ )); do
```

Ichimonji10 commented on rpm/gen-fixtures-delta.sh:
```diff
@@ -0,0 +1,154 @@
+# e.g. assets/walrus-5.21-1.noarch.rpm → 5.21
+get_rpm_version() {
+    local filename="$(basename "${1}")"
+    IFS=$'\n' local parts=($(echo "${filename}" | tr - "\n"))
```
I *just* barely learned about the value of declaring variables separately from using them. You can see an example of the dangers [here](http://pastebin.com/m9VPQ1KS). In short, write this instead:

```bash
    local filename parts
    filename="$(basename "${1}")"
    IFS=$'\n' parts=($(echo "${filename}" | tr - "\n"))
```

Ichimonji10 commented:
Many of my comments come from [shellcheck](https://www.shellcheck.net/). You can use shellcheck by installing it using your package manager and executing a command like `find . -name '*sh' -print0 | xargs -0 shellcheck`. This is, of course, a pain to type out. A better option is to rebase your development branch onto the upstream/master branch. Some [improvements](https://github.com/PulpQE/pulp-fixtures/commits/master) have been made, such as the addition of the `lint` make target.

Also, if you know how to squash commits, please do so. Otherwise, the reviewer who merges your work will do so. Beware that if the reviewer squashes your commits for you, they will be forced to come up with an original commit message, and it may or may not be to your liking.

Given the simple commit history at play, it's possible to rebase and squash your changes with the following script:

```bash
#!/usr/bin/env bash
set -euo pipefail

# Clone the existing repository and verify that it works.
git clone https://github.com/emmawu/pulp-fixtures.git
cd pulp-fixtures
git checkout develop
make fixtures/{drpm-unsigned,rpm-unsigned,srpm,srpm-unsigned}

# Add a remote tracking branch named "upstream".
git remote add upstream https://github.com/PulpQE/pulp-fixtures.git
git fetch --all

# Rebase the "develop" branch, and squash its commits.
git branch develop-old  # make a backup branch
git rebase upstream/master
git rebase --interactive upstream/master develop

# Verify that the updated code still works, and that the commit history looks
# sane.
make clean
make fixtures/{drpm-unsigned,rpm-unsigned,srpm,srpm-unsigned}
git log --decorate --graph --pretty=oneline --abbrev-commit --all

# If you're confident in the changes, push them.
# git push --force origin develop
```

By the way, notice how the `--force` option is used when pushing the branch? This is because we're *intentionally* throwing away commits. Git is very good about keeping commits around and warning you when commits may be lost.

Also, a big **thank you** for shrinking down the size of the binary blobs. I'm very, very happy to see that. :smile:

Finally, here's what you'll want to do when executing `git rebase --interactive upstream/master`. There's instructions right in the text editor, but you may appreciate the example.

![screenshot_2016-07-27_15-48-57](https://cloud.githubusercontent.com/assets/1012989/17189976/fae62268-5411-11e6-807c-ed19aed65bfe.png)

emmawu commented:
Thank you, @Ichimonji10. The instructions you've given for shellcheck and commit squashing is very helpful.
I have squashed my commits and reabsed it on upstream/master. And also updated my scripts based on all of your comments and shellcheck complaint.  
Please review.

Ichimonji10 commented:
My only significant suggestion is that the `createrepo` command use the `--deltas` option, not the `--delta` option. The man pages for createrepo and createrepo_c both make mention of `--deltas`, not `--delta`. I'll do this in a follow-up commit. I've tested out the change, and nothing changes. I assume that `--delta` is supported for legacy reasons.

Ichimonji10 commented:
Merged, with a follow-up commit of 8a003e048dfa814301a66734d3a022a972821fe1. Thank you!
